### PR TITLE
Refactoring Graph Warp Module

### LINKED ISSUE #329: Add tests for GWM
Graph Warp Module PR is merged, but tests are not written yet.
https://github.com/pfnet-research/chainer-chemistry/pull/321

 - add tests
 - add example tests
 - refactor class names GGNN_GWM --> GGNNGWM or
 - gwm support option in original graph conv network instead of adding each class.
 - refactor preprocessor to change behavior by option, instead of adding each class.

### PR DESCRIPTION
**Pull request by mottodora**
Resolve #329 
- [x] Add unit test of preprocesor
- [x] Refactor preprocessor
- [x] Add unit test of GWM model
- [x] Refactor GWM model

### CODE REVIEW

**Comment by mottodora** [comment]
I finish implementation. Please review.

File: chainer_chemistry/models/gwm.py [hunk 1]
```diff
@@ -5,11 +5,145 @@
...
+
+        if self.dropout_ratio > 0.0:
+            # TODO: fail backward test
+            z = functions.dropout(z, ratio=self.dropout_ratio)
+        z = self.activation(z)
+        merged = (1 - z) * h + z * g
+        return merged
+
+
+class SuperNodeTransmitterUnit(chainer.Chain):
```
Comment (corochann) [inline chainer_chemistry/models/gwm.py]: can you add docstring?

File: chainer_chemistry/models/gwm.py [hunk 2]
```diff
@@ -5,11 +5,145 @@
...
+        # intermediate_h_super.shape == (mb, self.hidden_dim)
+        g_trans = functions.tanh(g_trans)
+        # intermediate_h_super.shape == (mb, 1, self.hidden_dim)
+        g_trans = functions.expand_dims(g_trans, 1)
+        # intermediate_h_super.shape == (mb, atom, self.hidden_dim)
+        g_trans = functions.broadcast_to(g_trans, (mb, n_atoms, self.hidden_dim))
+        return g_trans
+
+
+class GraphTransmitterUnit(chainer.Chain):
```
Comment (corochann) [inline chainer_chemistry/models/gwm.py]: can you add docstring?

File: chainer_chemistry/models/gwm.py [hunk 3]
```diff
@@ -5,11 +5,145 @@
 from chainer_chemistry.links import GraphLinear
 
 
+class WarpGateUnit(chainer.Chain):
```
Comment (corochann) [inline chainer_chemistry/models/gwm.py]: can you add docstring?

File: chainer_chemistry/models/gwm.py [hunk 4]
```diff
@@ -5,11 +5,145 @@
...
+    def __init__(self, output_type='graph', hidden_dim=16,
+                 dropout_ratio=-1, activation=functions.sigmoid):
+        super(WarpGateUnit, self).__init__()
+        if output_type == 'graph':
+            LinearFunc = GraphLinear
+        elif output_type == 'super':
+            LinearFunc = links.Linear
+        else:
+            raise ValueError
```
Comment (corochann) [inline chainer_chemistry/models/gwm.py]: show proper error message.
`ValueError('output_type = {} is unexpected. graph or super is supported.'.format(output_type))`

**Comment by corochann** [comment]
Can you update code? @mottodora

**Comment by corochann** [comment]
I will take over.

File: tests/models_tests/test_gwm.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    embed_atom_data = cuda.to_gpu(embed_atom_data)
+    new_embed_atom_data = cuda.to_gpu(new_embed_atom_data)
+    supernode = cuda.to_gpu(supernode)
+    gwm.to_gpu()
+    check_forward(gwm, embed_atom_data, new_embed_atom_data, supernode)
+
+
+def check_backward(gwm, embed_atom_data, new_embed_atom_data, supernode,
+                   y_grad, supernode_grad):
+    gwm.GRU_local.reset_state()
```
[the remark this answers is not in the document]
Comment (corochann) [reply tests/models_tests/test_gwm.py]: done in new PR

File: tests/models_tests/test_gwm.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+def test_super_warp_gate_unit_backward(super_warp_gate_unit, data):
+    supernode = data[2]
+    supernode_grad = data[4]
+    gradient_check.check_backward(super_warp_gate_unit,
+                                  (supernode, supernode),
+                                  supernode_grad, eps=0.01)
+
+
+def check_forward(gwm, embed_atom_data, new_embed_atom_data, supernode):
+    gwm.GRU_local.reset_state()
```
[the remark this answers is not in the document]
Comment (corochann) [reply tests/models_tests/test_gwm.py]: done in new PR

File: chainer_chemistry/models/gwm.py [hunk 7]
```diff
@@ -106,124 +225,37 @@ def __call__(self, h, h_new, g, step=0):
                 current local node hidden states as input of the vanilla GNN
         :param h_new: minibatch by num_nodes by hidden_dim numpy array.
                 updated local node hidden states as output from the vanilla GNN
-        :param adj: minibatch by bond_types by num_nodes by num_nodes 1/0 array.
-                Adjacency matrices over several bond types
+        :param adj: minibatch by bond_types by num_nodes by num_nodes 1/0
+                array. Adjacency matrices over several bond types
```
[the remark this answers is not in the document]
Comment (corochann) [reply chainer_chemistry/models/gwm.py]: done in new PR

File: chainer_chemistry/models/gwm.py [hunk 8]
```diff
@@ -22,80 +156,65 @@ class GWM(chainer.Chain):
...
         if tying_flag:
-            num_layer = 1
+            n_layers = 1
 
         with self.init_scope():
+            self.update_super = chainer.ChainList(
+                *[links.Linear(in_size=hidden_dim_super,
+                               out_size=hidden_dim_super)
+                  for _ in range(n_layers)]
+            )
```
[the remark this answers is not in the document]
Comment (corochann) [reply chainer_chemistry/models/gwm.py]: It is different from original code, but separating F_super is better and it's ok.

File: chainer_chemistry/dataset/preprocessors/common.py [hunk 9]
```diff
@@ -169,7 +169,40 @@ def construct_discrete_edge_matrix(mol, out_size=-1):
         adjs[ch, j, i] = 1.0
     return adjs
 
-def construct_supernode_feature(mol, atom_array, adjs, largest_atomic_number=MAX_ATOMIC_NUM, out_size=-1):
+
+def mol_basic_info_feature(mol, atom_array, adj):
+    n_atoms = mol.GetNumAtoms()
+    assert n_atoms == len(atom_array)
+    n_edges = adj.sum()
```
[the remark this answers is not in the document]
Comment (corochann) [reply chainer_chemistry/dataset/preprocessors/common.py]: done in new PR

File: chainer_chemistry/dataset/preprocessors/common.py [hunk 10]
```diff
@@ -169,7 +169,40 @@ def construct_discrete_edge_matrix(mol, out_size=-1):
...
+
+def mol_bond_freq_feature(mol, atom_array, adj):
+    if adj.ndim == 2:
+        adj = numpy.expand_dims(adj, axis=0)
+    adj = adj.reshape((adj.shape[0], -1))
+    return adj.sum(axis=1) / adj.sum()
+
+
+def construct_supernode_feature(mol, atom_array, adj, feature_functions=None):
+                                # largest_atomic_number=MAX_ATOMIC_NUM, out_size=-1):
```
[the remark this answers is not in the document]
Comment (corochann) [reply chainer_chemistry/dataset/preprocessors/common.py]: done in new PR

File: chainer_chemistry/models/gwm.py (hunk 3 above)
Comment (corochann) [reply chainer_chemistry/models/gwm.py]: done in new PR

File: chainer_chemistry/models/gwm.py (hunk 4 above)
Comment (corochann) [reply chainer_chemistry/models/gwm.py]: done in new PR

File: chainer_chemistry/models/gwm.py [hunk 11]
```diff
@@ -5,11 +5,145 @@
...
+            self.H = LinearFunc(in_size=hidden_dim, out_size=hidden_dim)
+            self.G = LinearFunc(in_size=hidden_dim, out_size=hidden_dim)
+
+        self.hidden_dim = hidden_dim
+        self.dropout_ratio = dropout_ratio
+        self.output_type = output_type
+        self.activation = activation
+
+    def __call__(self, h, g):
+        z = self.H(h) + self.G(g)
```
[the remark this answers is not in the document]
Comment (corochann) [reply chainer_chemistry/models/gwm.py]: TODO later

File: chainer_chemistry/models/gwm.py (hunk 1 above)
Comment (corochann) [reply chainer_chemistry/models/gwm.py]: done in new PR

File: chainer_chemistry/models/gwm.py [hunk 12]
```diff
@@ -5,11 +5,145 @@
...
+        merged = (1 - z) * h + z * g
+        return merged
+
+
+class SuperNodeTransmitterUnit(chainer.Chain):
+    def __init__(self, hidden_dim_super=16, hidden_dim=16, dropout_ratio=-1):
+        super(SuperNodeTransmitterUnit, self).__init__()
+        with self.init_scope():
+            self.F_super = links.Linear(in_size=hidden_dim_super,
+                                        out_size=hidden_dim)
```
[the remark this answers is not in the document]
Comment (corochann) [reply chainer_chemistry/models/gwm.py]: F_super is separated and I think it's okay.

File: chainer_chemistry/models/gwm.py [hunk 13]
```diff
@@ -5,11 +5,145 @@
...
+        self.hidden_dim = hidden_dim
+        self.hidden_dim_super = hidden_dim_super
+        self.dropout_ratio = dropout_ratio
+
+    def __call__(self, g, n_atoms):
+        mb = len(g)
+        # for local updates
+        g_trans = self.F_super(g)
+        # intermediate_h_super.shape == (mb, self.hidden_dim)
+        g_trans = functions.tanh(g_trans)
```
[the remark this answers is not in the document]
Comment (corochann) [reply chainer_chemistry/models/gwm.py]: TODO later

File: chainer_chemistry/models/gwm.py (hunk 2 above)
Comment (corochann) [reply chainer_chemistry/models/gwm.py]: done in new PR

File: chainer_chemistry/models/gwm.py [hunk 14]
```diff
@@ -5,11 +5,145 @@
...
+
+
+class GraphTransmitterUnit(chainer.Chain):
+    def __init__(self, hidden_dim_super=16, hidden_dim=16, n_heads=8,
+                 dropout_ratio=-1, activation=functions.tanh):
+        super(GraphTransmitterUnit, self).__init__()
+        with self.init_scope():
+            self.V_super = links.Linear(hidden_dim * n_heads, hidden_dim * n_heads)
+            self.W_super = links.Linear(hidden_dim * n_heads, hidden_dim_super)
+            self.B = GraphLinear(n_heads * hidden_dim, n_heads * hidden_dim_super)
```
[the remark this answers is not in the document]
Comment (corochann) [reply chainer_chemistry/models/gwm.py]: done in new PR

File: chainer_chemistry/models/gwm.py [hunk 15]
```diff
@@ -5,11 +5,145 @@
...
+        # g_extend.shape == (mb, self.n_heads, self.hidden_dim_super)
+        g_extend = functions.broadcast_to(g_extend, (mb, self.n_heads,
+                                                     self.hidden_dim_super))
+        # g_extend.shape == (mb, self.n_heads, 1, self.hidden_dim_super)
+        g_extend = functions.expand_dims(g_extend, 2)
+
+        # update for attention-message B h_i
+        # h1.shape == (mb, atom, n_heads * ch)
+        # Bh_i.shape == (mb, atom, self.n_heads * self.hidden_dim_super)
+        Bh_i = self.B(h1)
```
[the remark this answers is not in the document]
Comment (corochann) [reply chainer_chemistry/models/gwm.py]: done in new PR

File: chainer_chemistry/models/gwm.py [hunk 16]
```diff
@@ -5,11 +5,145 @@
...
+        attention_sum = functions.matmul(attention_i, h_j)
+        # attention_sum.shape == (mb, self.n_heads * ch)
+        attention_sum = functions.reshape(attention_sum,
+                                          (mb, self.n_heads * ch))
+
+        # weighting h for different heads
+        # intermediate_h.shape == (mb, self.n_heads * ch)
+        h_trans = self.V_super(attention_sum)
+        # compress heads
+        h_trans = self.W_super(h_trans)
```
[the remark this answers is not in the document]
Comment (corochann) [reply chainer_chemistry/models/gwm.py]: done in new PR

File: chainer_chemistry/models/gwm.py [hunk 17]
```diff
@@ -5,11 +5,145 @@
...
+        # h1.shape == (mb, atom, 1, ch)
+        h1 = functions.expand_dims(h, 2)
+        # h1.shape == (mb, atom, n_heads, ch)
+        h1 = functions.broadcast_to(h1, [mb, atom, self.n_heads, ch])
+        # h1.shape == (mb, atom, n_heads * ch)
+        h1 = functions.reshape(h1, [mb, atom, self.n_heads * ch])
+
+        h_j = functions.expand_dims(h, 1)
+        # h_j.shape == (mb, self.n_heads, atom, ch)
+        h_j = functions.broadcast_to(h_j, (mb, self.n_heads, atom, ch))
```
Comment (corochann) [inline chainer_chemistry/models/gwm.py]: apply V_super instead of broadcast

Comment (corochann) [reply chainer_chemistry/models/gwm.py]: TODO later

File: chainer_chemistry/models/gwm.py [hunk 18]
```diff
@@ -22,80 +156,65 @@ class GWM(chainer.Chain):
             number of super-node observation attributes
         n_edge_types (int): number of edge types witin graphs.
         dropout_ratio (default=0.5); if > 0.0, perform dropout
-        tying_flag (default=false): enable if you want to share params across layers
+        tying_flag (default=false): enable if you want to share params across
+            layers
```
[the remark this answers is not in the document]
Comment (corochann) [reply chainer_chemistry/models/gwm.py]: done in new PR

File: chainer_chemistry/models/gwm.py [hunk 19]
```diff
@@ -22,80 +156,65 @@ class GWM(chainer.Chain):
...
+    def __init__(self, hidden_dim=16, hidden_dim_super=16, n_layers=4,
+                 n_heads=8, dropout_ratio=-1, concat_hidden=False,
+                 tying_flag=False, activation=functions.relu,
+                 wgu_activation=functions.sigmoid,
+                 gtu_activation=functions.tanh):
         super(GWM, self).__init__()
-        num_layer = n_layers
         if tying_flag:
-            num_layer = 1
+            n_layers = 1
```
[the remark this answers is not in the document]
Comment (corochann) [reply chainer_chemistry/models/gwm.py]: Instead, refactor to remove n_layers dependency and do not use `step` in __call__ method.
TODO later.

**Pull request merged**